// mackerel-cloud-integration/src/mackerel/api/model/alert.ts
/**
 * Makerel API
 * This is part of Makerel API.
 *
 * The version of the OpenAPI document: 0.1.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


export class Alert {
    'id': string;
    'status': string;
    'monitorId': string;
    'type': string;
    'hostId'?: string;
    'value'?: number;
    'message'?: string;
    'reason'?: string;
    'openedAt': number;
    'closedAt'?: number;

    static discriminator: string | undefined = undefined;

    static attributeTypeMap: Array<{name: string, baseName: string, type: string}> = [
        {
            "name": "id",
            "baseName": "id",
            "type": "string"
        },
        {
            "name": "status",
            "baseName": "status",
            "type": "string"
        },
        {
            "name": "monitorId",
            "baseName": "monitorId",
            "type": "string"
        },
        {
            "name": "type",
            "baseName": "type",
            "type": "string"
        },
        {
            "name": "hostId",
            "baseName": "hostId",
            "type": "string"
        },
        {
            "name": "value",
            "baseName": "value",
            "type": "number"
        },
        {
            "name": "message",
            "baseName": "message",
            "type": "string"
        },
        {
            "name": "reason",
            "baseName": "reason",
            "type": "string"
        },
        {
            "name": "openedAt",
            "baseName": "openedAt",
            "type": "number"
        },
        {
            "name": "closedAt",
            "baseName": "closedAt",
            "type": "number"
        }    ];

    static getAttributeTypeMap() {
        return Alert.attributeTypeMap;
    }
}
