// mackerel-cloud-integration/lib/mackerel/api/model/hostStatusRequest.js
"use strict";
/**
 * Makerel API
 * This is part of Makerel API.
 *
 * The version of the OpenAPI document: 0.1.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var HostStatusRequest = /** @class */ (function () {
    function HostStatusRequest() {
    }
    HostStatusRequest.getAttributeTypeMap = function () {
        return HostStatusRequest.attributeTypeMap;
    };
    HostStatusRequest.discriminator = undefined;
    HostStatusRequest.attributeTypeMap = [
        {
            "name": "status",
            "baseName": "status",
            "type": "HostStatusRequest.StatusEnum"
        }
    ];
    return HostStatusRequest;
}());
exports.HostStatusRequest = HostStatusRequest;
(function (HostStatusRequest) {
    var StatusEnum;
    (function (StatusEnum) {
        StatusEnum[StatusEnum["Standby"] = 'standby'] = "Standby";
        StatusEnum[StatusEnum["Working"] = 'working'] = "Working";
        StatusEnum[StatusEnum["Maintenance"] = 'maintenance'] = "Maintenance";
        StatusEnum[StatusEnum["Poweroff"] = 'poweroff'] = "Poweroff";
    })(StatusEnum = HostStatusRequest.StatusEnum || (HostStatusRequest.StatusEnum = {}));
})(HostStatusRequest = exports.HostStatusRequest || (exports.HostStatusRequest = {}));
exports.HostStatusRequest = HostStatusRequest;
